### ISSUE TITLE
Incorrect order of extracted highlights

### ISSUE THREAD

**Issue by ichnilatis-gr** [koreader/koreader#8601]
* KOReader version: v2021.12.1
* Device: PB 628

#### Issue
Some times the extracted highlights don't follow the order of the highlights in the book. You can see this in the following screenshots. (Sorry that they are from a Greek book).

![Reader_2021-12-29_100018](https://user-images.githubusercontent.com/76203123/147679266-cc401134-f7c0-45e6-acbb-c9b518a8fe38.jpg)

![Reader_2021-12-29_095839](https://user-images.githubusercontent.com/76203123/147679275-8d1008de-b291-4593-bd73-0736c39bb83b.jpg)

**Comment by hius07** [comment]
I guess the bookmark No 2 was added first on this page? (Check time)
Within one page bookmarks are sorted in order of adding them.
I agree that this should be fixed.

**Comment by ichnilatis-gr** [comment]
The highlights were made in the order they appear on the page (as they are numbered in the first screenshot).

**Comment by hius07** [comment]
Can you share a metadata.*.lua file of this book (it is in the <bookname>.sdr folder that is near the book).

**Comment by ichnilatis-gr** [comment]
[Uploading metadata.pdf.lua.txt…]()

**Comment by ichnilatis-gr** [comment]
It's 1.26 MB and it seems difficult to be uploaded... :-)

**Comment by Frenzie** [comment]
The limit should be something like 10 MB. Maybe if you zip it?

**Comment by ichnilatis-gr** [comment]
[metadata.html - Copy.lua.zip](https://github.com/koreader/koreader/files/7802644/metadata.html.-.Copy.lua.zip)

**Comment by hius07** [comment]
There are no bookmarks in that file.

**Comment by ichnilatis-gr** [comment]
Excuse me, my false. Wrong file.

[metadata.pdf - Copy.lua.zip](https://github.com/koreader/koreader/files/7802787/metadata.pdf.-.Copy.lua.zip)

**Comment by hius07** [comment]
Thank you for the file. I see that the bookmarks of page 261 are in good order.
I guess they are displayed in the Bookmark list in good order too.

It seems that the issue occurs in the Exporter plugin that tries to finally sort the list to be exported
https://github.com/koreader/koreader/blob/6c9e222d332f039972d4e7aabc2b11d9247a6f7b/plugins/exporter.koplugin/clip.lua#L286
taking into account page numbers only.
When page numbers of several elements are the same, we cannot get the correct order of the bookmarks of this page.

I'm now working on fixing the bookmark module to get bookmarks of one page correctly sorted depending on their position in the text (now they are sorted in order of adding).
If I succeed (hope so) I'll try to look into the plugin more closely, it would be easier to fix it based on the well sorted bookmarks list.

**Comment by ichnilatis-gr** [comment]
For some reason, after having installed the last nightly versions (v2021.12.1-29 and v2021.12.1-41), highlights are not extracted from the book given above as example. The process takes place, but I cannot find the file in the clipboard folder. This problem doesn't exist for other pdf files.

Any idea?

**Comment by Frenzie** [comment]
@NiLuJe mentioned a conversion might still be missing in current nightlies or something?

**Comment by NiLuJe** [comment]
That was an opds server added to the default list, IIRC, nothing to do with this ;).

**Comment by poire-z** [comment]
Fixed by #8616.

**Comment by ichnilatis-gr** [comment]
I thought that this issue had been resolved, but today I noticed that it hasn't.
I give you (unfortunately for you in Greek) a page containing highlights and their view in the extracted file.
The order of the extracted highlights is the reverse of the order in the book.

![Reader_2022-04-09_140248](https://user-images.githubusercontent.com/76203123/162574371-f8ff231c-f618-4f4c-8e4b-b6e54630cd8d.jpg)
![Reader_2022-04-09_140533](https://user-images.githubusercontent.com/76203123/162574374-23ab1430-8bfc-4e86-8cba-be9d2a17251e.jpg)
